cpp/src/arrow/error_or.h

```diff
@@ -0,0 +1,427 @@
+  ErrorOr& operator=(const ErrorOr& other) {
+    // Check for self-assignment.
+    if (this == &other) {
+      return *this;
+    }
+
+    // Construct the variant object using the variant object of the source.
+    if (other.has_value_) {
+      AssignValue(other.variant_.value_);
+    } else {
+      AssignStatus(other.variant_.status_);
+    }
+    return *this;
+  }
+
+  /// Templatized constructor which constructs a `ErrorOr<T>` by moving the
+  /// contents of a `ErrorOr<U>`. `T` must be implicitly constructible from `U
+  /// &&`.
+  ///
+  /// Sets `other` to contain a non-OK status with a`StatusError::Invalid`
+  /// error code.
+  ///
+  /// \param other The ErrorOr object to move from and set to a non-OK status.
+  template <typename U,
+            typename E = typename std::enable_if<std::is_constructible<T, U>::value &&
+                                                 std::is_convertible<U, T>::value>::type>
+  ErrorOr(ErrorOr<U>&& other) : has_value_(other.has_value_) {
+    if (has_value_) {
+      new (&variant_) variant(std::move(other.variant_.value_));
+      other.OverwriteValueWithStatus(
+          Status::Invalid(ErrorOrConstants::kValueMoveConstructorMsg));
+    } else {
+      new (&variant_) variant(std::move(other.variant_.status_));
+#ifndef NDEBUG
+      // The other.variant_.status_ gets moved and invalidated with a Status-
+      // specific error message above. To aid debugging, set the status to a
+      // ErrorOr-specific error message.
+      other.variant_.status_ =
+          Status::Invalid(ErrorOrConstants::kStatusMoveConstructorMsg);
+#endif
+    }
+  }
+
+  /// Move-assignment operator.
+  ///
+  /// Sets `other` to contain a non-OK status.
+  ///
+  /// \param other The ErrorOr object to assign from and set to a non-OK
+  /// status.
+  ErrorOr& operator=(ErrorOr&& other) {
+    // Check for self-assignment.
+    if (this == &other) {
+      return *this;
+    }
+
+    // Construct the variant object using the variant object of the donor.
+    if (other.has_value_) {
+      AssignValue(std::move(other.variant_.value_));
+      other.OverwriteValueWithStatus(
+          Status::Invalid(ErrorOrConstants::kValueMoveAssignmentMsg));
+    } else {
+      AssignStatus(std::move(other.variant_.status_));
+#ifndef NDEBUG
+      // The other.variant_.status_ gets moved and invalidated with a Status-
+      // specific error message above. To aid debugging, set the status to a
+      // ErrorOr-specific error message.
+      other.variant_.status_ =
+          Status::Invalid(ErrorOrConstants::kStatusMoveAssignmentMsg);
+#endif
+    }
+
+    return *this;
+  }
+
+  /// Indicates whether the object contains a `T` value.
+  ///
+  /// \return True if this ErrorOr object's status is OK (i.e. a call to ok()
+  /// returns true). If this function returns true, then it is safe to access
+  /// the wrapped element through a call to ValueOrDie().
+  bool ok() const { return has_value_; }
```

Comment:
Are you suggesting adding one?  I added a comment recommending using ASSIGN_OR_RAISE.  But can add an operator as well if you think it is worthwhile.

Reply:
I think it's worth adding the bool operator.

Reply:
What do you think of:
```c++
Status TryAssign(U* out) && {
  if (ok()) {
    *out = ValueOrDie();
  }
  return status();
}
```

This would allow usage like:
```c++
int i;
RETURN_NOT_OK(maybe_int().TryAssign(&i));
```

Which could be further simplified with a more specialized macro:
```c++
#define ARROW_VALUE_OR_RAISE(var, rexpr) \
  decltype((rexpr).ValueOrDie()) var;  \
  ARROW_RETURN_NOT_OK((rexpr).TryAssign(&var))
```

Reply:
@fsaintjacques adding bool breaks test cases, due to being castable to an integer.  I suppose there is some level of template magic that might make it work but I'm not sure it is worth it.  Let me know if you want me to try harder.

@bkietz this is closer to how the original macro worked.  I think it is cleaner to be able to have 
```
ARROW_ASSIGN_OR_RAISE(int i, SomeFunction());
```
I think it detracts a little bit from the value of having to declare variables on a second line because it looks very similar to the pattern today we have with output parameters.

Reply:
Which test case breaks?